### ISSUE TITLE
subdatasets() (and others) reporting issue

### ISSUE THREAD

**Issue by mih** [datalad/datalad#4413]
I am in a full checkout of the HCP dataset, and I run:

```
% datalad subdataset -d . -r
subdataset(ok): HCP1200/100206 (dataset)
subdataset(ok): HCP1200/100206/MNINonLinear (dataset)
subdataset(ok): HCP1200/100206/T1w (dataset)
subdataset(ok): HCP1200/100206/unprocessed (dataset)
subdataset(ok): HCP1200/100307 (dataset)
subdataset(ok): HCP1200/100307/MEG (dataset)
subdataset(ok): HCP1200/100307/MNINonLinear (dataset)
subdataset(ok): HCP1200/100307/T1w (dataset)
subdataset(ok): HCP1200/100307/unprocessed (dataset)
subdataset(ok): HCP1200/100408 (dataset)
```

I get the output above almost immediately, but then nothing else happens. Looking with `-l debug` DataLad is very busy though:

```
...
[DEBUG  ] <AnnexRepo path=/ds/hcp/human-connectome-project-openaccess/HCP1200/100408 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
[DEBUG  ] Done query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
[DEBUG  ] Done <AnnexRepo path=/ds/hcp/human-connectome-project-openaccess/HCP1200/100408 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] <AnnexRepo path=/ds/hcp/human-connectome-project-openaccess/HCP1200/100610 (<class 'datalad.support.annexrepo.AnnexRepo'>)>.get_content_info(...) 
[DEBUG  ] Query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
[DEBUG  ] Done query repo: ['git', 'ls-files', '--stage', '-z', '-d', '-m', '--exclude-standard'] 
...
```

but not reporting anything. It eventually completes and reports all results at once.

This smells bad and is not what should happen.

**Comment by mih** [comment]
Oi! This might all be an artifact of the recently introduces result suppression!

**Comment by kyleam** [comment]
Hrm, tricky... Perhaps we should whitelist the suppression to particular actions we think are speedy?

Another option could be to have some time counter for the suppression (perhaps total and per item) and then jump out of the suppression, as we would if we hit a "different" result.  In that case, we'd essentially hop back in and out of suppression, giving the user some indication of what's going on.  Not sure how that'd be from an experience standpoint.

**Comment by mih** [comment]
> Another option could be to have some time counter for the suppression (perhaps total and per item) and then jump out of the suppression, as we would if we hit a "different" result. In that case, we'd essentially hop back in and out of suppression, giving the user some indication of what's going on. Not sure how that'd be from an experience standpoint.

I like that! IOW, we could ensure that a (single?) result comes through at least every X seconds (maybe 10s?).

**Comment by kyleam** [comment]
> maybe 10s?

Sounds sensible.

Would you like me to work on a PR for that today, or are you on it?

**Comment by mih** [comment]
> Would you like me to work on a PR for that today, or are you on it?

I cannot work on it immediately, and I fucked up the last one. So if you feel the energy, I'd love if you could give it a go!

**Comment by yarikoptic** [comment]
FWIW, for some reason for me it is fast:
```
(git)smaug:/mnt/btrfs/datasets/datalad/crawl/hcp-openaccess[master]git
$> time datalad subdatasets      
subdataset(ok): HCP1200/100206 (dataset)
subdataset(ok): HCP1200/100307 (dataset)
subdataset(ok): HCP1200/100408 (dataset)
subdataset(ok): HCP1200/100610 (dataset)
subdataset(ok): HCP1200/101006 (dataset)
subdataset(ok): HCP1200/101107 (dataset)
subdataset(ok): HCP1200/101309 (dataset)
subdataset(ok): HCP1200/101410 (dataset)
subdataset(ok): HCP1200/101915 (dataset)
subdataset(ok): HCP1200/102008 (dataset)
  [1103 similar messages have been suppressed]
action summary:
  subdataset (ok: 1113)
datalad subdatasets  0.62s user 0.21s system 102% cpu 0.806 total
```

Note that behavior of producing output or not based on timing might result in tricky cases such as varying outputs of the same command depending on interaction between speed of the system and "speediness" of the command. Might negatively affect e.g. embedding traces of the commands into documentation (might always or frequently change output) and comparative analysis of the runs e.g. for debugging etc.
Related: just filed #4416 while thinking to discuss here that may be some commands should not even have their output suppressed.  But then I liked that my terminal didn't get flooded with all subdatasets of HCP.

**Comment by mih** [comment]
Add `--recursive` ;-)

**Comment by yarikoptic** [comment]
> Add `--recursive` ;-)

d'oh -- should have just cut/pasted ;) thanks

**Comment by yarikoptic** [comment]
ah, gotcha -- so it is waiting because it has no clue about how many more lines are yet to be suppressed while suppressing them. Well, could as well, in the interactive case, add a progress counter (not bar - since we have no clue how many to expect) there which would disappear at the end (like in other places with progress bar) and be replaced with the final message. In the non-interactive case - I guess ok as is ;)

**Comment by yarikoptic** [comment]
then it would be up to progress bar/counter to consistently (with other places) frequently update user about the "progress" (instead of coming up with some specific strategy in renderer)

**Comment by yarikoptic** [comment]
progress counter here could be as nice as just the same message but with growing number, which at the end is replaced with the final one ;)

**Comment by kyleam** [comment]
> > Would you like me to work on a PR for that today, or are you on it?
> 
> I cannot work on it immediately, and I fucked up the last one. So if you feel the energy, I'd love if you could give it a go!

Seems there has been a flurry of other ideas, some directly incompatible with mine.  Sorry, I'm not volunteering for any of those :/

**Comment by mih** [comment]
I think the original idea would very much work like a progress "counter". It would be minimal change from previous and current behavior, and require a rather localized patch. Moreover, I believe this is something that should be addressed prior 0.13 and having @kyleam volunteer means that it will get done. We can always have something better later, with a new volunteer (that is also not me ;-)
